Add documentation for config-validator module

Kenny407 opened the pull request:
This PR aims to add a clear documentation of the different parts a deployment file can contain with examples and usages, the current supported fields and the notes/types required for each section.

Please don't hesitate to add comments whether of wording or if something is not clear. 

Thanks 🙏

Olivier-Daire commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,163 @@
+
+```yaml
+processes:
+  - name: process-name  # Process name, STRING - Required
````
I believe that we have a limit on the process name length, it could be useful to document it here

Olivier-Daire commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,163 @@
+
+```yaml
+processes:
+  - name: process-name  # Process name, STRING - Required
+    is_cronjob: false   # Defining a simple process, BOOLEAN - Required
+    start_command: npm start  # Executed command when the pod is started. STRING - Required
+  - name: cronjob-name  # Process name - Required
+    is_cronjob: true    # Defining a scheduled process (cronjob) - Required
+    start_command: npm run worker # Executed command when the pod is started - Required
````
```suggestion
# Defining a simple process
  - name: process-name  # STRING - Required
    is_cronjob: false  # BOOLEAN - Required
    start_command: npm start  # Executed command when the pod is started. STRING - Required
# Defining a scheduled process
    - name: cronjob-name
    is_cronjob: true
    start_command: npm run worker
```

benard-g commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,163 @@
+This is a basic configuration application with the minimum required fields.
+
+```yaml
+app: my-application   # Application's name
+is_enabled: false     # Is application deployment enabled
+git: git@microservice_repo # Git repository URL
+variables:            # Environment variables
````
I would try to align the comments (same for all comments)
```suggestion
app: my-application        # Application's name
is_enabled: false          # Is application deployment enabled
git: git@microservice_repo # Git repository URL
variables:                 # Environment variables
```

benard-g replied:
Also, you sometimes end up with a lot of information in some of the comments (not especially here but in the other examples too).
What do you think of using a table describing what is expected and then having the example ? 🙂

Olivier-Daire commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,163 @@
+```
+
+### Dependencies
+The list of dependencies used by your project.
````
Even as a Kapten developer, I am not sure what this list is used for. We should either remove it, if there is no use, or dig a bit more to find out why we need this and document it more.

benard-g replied:
What are dependencies exactly ? some other apps defined in the folder ?

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,163 @@
+  my-process:         # Process to scale
+    min-replicas: 1   # Minimal number of pods, INTEGER - Required
+    max-replicas: 10  # Minimal number of pos, INTEGER - Required
+    targetCPUUtilizationPercentage: # Percentage of CPU which will trigger a horizontal scaling when it is exceeded, INTEGER, 0 <= x <= 100 - Required
```
Is it possible to rename it or is this a fixed name ?
```suggestion
    targetCpuUtilizationPercentage: # Percentage of CPU which will trigger a horizontal scaling when it is exceeded, INTEGER, 0 <= x <= 100 - Required
```

benard-g replied:
Also, add a value before the comment to make the block a valid yaml (e.g. 95)

Kenny407 replied:
I think it would be a good time if we are going to keep the naming for acronyms in variables for example: HTML vs Html, CPU vs Cpu, and so on. In my opinion I like better to keep the acronyms in uppercase as long as they are not at the beginning of the variable definition.

Olivier-Daire commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,163 @@
+```
+
+### Public
+Determines if the project is publicly available or not 
````
Public how? I guess there is some Ingress logic behind this, how is it linked to a domain name? I think we should detail this part

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,163 @@
+dependencies:
+  - first-dependency
+  - second-dependency
+  ...
```
I would suggest to use a comment for this kind of "etc" statement to keep the `yaml` valid when copy/paste
```suggestion
  # ...
```

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,163 @@
+
+### Probes
+
+If you need more documentation on what probes are, visit the [official documentation](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes).
```
```suggestion
If you need more documentation on what probes are, visit the [official kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes).
```

Olivier-Daire replied:
👍 Good idea! 
Maybe put it in italic and at the end of the section?

Olivier-Daire commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,163 @@
+Define aliases for public access.
+
+```yaml
+public_aliases:
````
Same here, I have no clue how it works, in combination with `public` rule above

Olivier-Daire replied:
What about these?

Olivier-Daire commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,163 @@
+
+If you need more documentation on what probes are, visit the [official documentation](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes).
+
+In this section, you can define some probes Kubernetes for specific containers. Here is the only available probe for the moment:
```
```suggestion
You can define some Kubernetes' probes for each process.
```
I we cannot define probes for other processes than web, maybe change it to something like   `You can define probes for the web process. It is currently not possible to define probes for other processes.`

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,163 @@
+
+### Variables
+
+Environment variables used in the application.
```
It could be good to specify what are the categories `app`/`ope`/... 
Or maybe this should rather be addressed in nestor-api's doc 🤷 
What do you think ?

Kenny407 replied:
Done 👍

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,163 @@
+    APP_VAR1: application_variable_1
+  ope: # Required
+    OPE_VAR1: ope_variable_1
+  integration: # The unique env vars (which won't be replaced be staging env vars) belonging to the integration, Optional
```
I was not aware of this mechanism 🤔 
Does this mean that it protects the variables in the replicas from being removed/overridden when staging is copied ?

Kenny407 replied:
It was supported before but it was prone to errors from the user part. I will remove it and just keep ope and env

Olivier-Daire commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,163 @@
+    APP_VAR1: application_variable_1
+  ope: # Required
+    OPE_VAR1: ope_variable_1
+  integration: # The unique env vars (which won't be replaced be staging env vars) belonging to the integration, Optional
```
I am not sure we should keep this feature? Seems a bit weird to me, and having a configuration per environment is better, imo, what do you think?

Kenny407 replied:
Removed! 👍

benard-g commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,163 @@
+    OPE_VAR1: ope_variable_1
+```
+
+To customize your application deployment, there are other sections supported. Check the following sections.
````
I am not sure if it is clearer, what do you tink ?
```suggestion
To further customize the deployment of your application, you can add the following sections to the configuration files : 
```

Olivier-Daire commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,163 @@
+### Crons
+Crons are tasks that run at a specific time or interval. You can use this type of features to automate tasks, commonly as: reporting, backups, maintenance, emails and so on.
+
+You can define the crons information with the following structure.
```
```suggestion
You can define the crons configuration with the following structure.
```

Olivier-Daire commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,163 @@
+  another-process: ...
+```
+
+### Crons
````
What do you think about putting this section right after `processes`? Since we already talk about cron in the processes part

Olivier-Daire commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,163 @@
+```
+
+
+### TemplateVars
````
I am not sure what these variables are used for, I am wondering if we should not omit them first, to keep it simple and see afterwards if we need them?

Kenny407 replied:
These variables from my understanding are used to override default values for Kubernetes. I will add the explanation in the introductory paragraph so it's clearer 👍

Kenny407 commented:
@Olivier-Daire @benard-g I improved the descriptions, re-arranged some sections, explained the usage that might not have been clear before. And added more meaningful examples. Also removed a few things that might not be needed for now. 🙏 thanks in advance for the re-review

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,195 @@
+
+In each file you will define the application that you want to deploy, few fields are required to setup a new application. There are many supported values which are optional that could be useful for your deployment. You can read more about them in the following sections and the schemas definition.
+
+<a name="applicationConfiguration"></a>
```
This tag a second time ? (it is also present on line 11)

benard-g commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,195 @@
+```yaml
+crons:
+  my-cronjob: # Cronjob to define, it must be the same name as the cronjob defined in processes
+    concurrency_policy: 'Forbid' # Allow, Forbid or Replace concurrency of cronjobs, STRING - Required. [Kubernetes documentation](https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/#concurrency-policy)
````
This line is really long, put the link after the block as a note to keep the comments simple.
And align the other comments together 🙂

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,195 @@
+### Scales
+When you deploy an application in a Kubernetes environment, you define how many replicas you would like to run. 
+Each replica represents a Kubernetes Pod that encapsulates your application container. 
+And also you can increase, or reduce the number of replicas in the given limit.
```
```suggestion
You can also increase, or reduce the number of replicas in the given limit.
```

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,195 @@
+
+In this section you can define variables used in configuration templates when building the kubernetes configuration before deployment. Here are the available variables:
+
+**_Note_**: All the variables should be prefixed by `tpl`. To avoid colliding with the 
```
`with the ...` ? 🙂

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,195 @@
+
+### Dependencies
+
+This section can be used when your application depends or makes use of another applications. In this section you can specify the list dependencies used by your project. 
```
```suggestion
This section can be used when your application depends or makes use of other applications. In this section you can specify the list dependencies used by your project. 
```

benard-g commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,195 @@
+- `app`: These are the variables used internally for adding logic to the application, such as: fixed values, timeouts, default values, feature flags, ids, and more.
+
+
+Environment variables used in the application.
```
This sentence look weird (and alone 😢). Did you perhaps forget to remove it after writing the paragraph above ?

Olivier-Daire commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,192 @@
+  tplCanary:          # Define name for canary, STRING
+```
+
+### Dependencies
````
I am not sure this part is useful and if we keep it, it should not be under Kubernetes, I think

Olivier-Daire commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,192 @@
+```
+
+### Public Aliases
+In this section allows at pod-level to override the hostname resolution when DNS and other options are not applicable. This is a feature of Kubernetes, that allows to reach your application with a more friendly name.
````
```suggestion 
This section allows at pod-level to override the hostname resolution when DNS and other options are not applicable. This is a feature of Kubernetes, that allows to reach your application with a more friendly name.
```

Olivier-Daire commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,192 @@
+### Public Aliases
+In this section allows at pod-level to override the hostname resolution when DNS and other options are not applicable. This is a feature of Kubernetes, that allows to reach your application with a more friendly name.
+
+You can define an alias for your application to facilitate public access.
```
```suggestion
You can define an alias for your application to ease public access.
```

Olivier-Daire commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,192 @@
+
+You can define an alias for your application to facilitate public access.
+
+If you need further documentation of how DNS and the network layer works in Kubernets. We recommend you to read [the official Kubernetes documentation](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/)
```
```suggestion
If you need further documentation about how DNS and the network layer works in Kubernetes. We recommend you to read [the official Kubernetes documentation](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/)
```

Olivier-Daire commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,192 @@
+### Variables
+There are two categories of environment variables that you can add to your application: `ope` and `app`. 
+
+The idea is to separate concerns between the variables that your application needs.
+
+- `ope`: These are used to specify operational values such as URLs, certificates, auth public keys, connection strings, integration with third party services and other services information used by your app.
+- `app`: These are the variables used internally for adding logic to the application, such as: fixed values, timeouts, default values, feature flags, ids, and more.
```
👍

Olivier-Daire commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,193 @@
+### Processes
+
+A project can contain multiple processes such as workers, web processes and more.
+The structure to define a process is the following one:
```
```suggestion
```

Olivier-Daire commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,193 @@
+
+**_Note_**: You can read more about how to write a cron, the different concurrency policies and more features in the [official documentation](https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/#writing-a-cron-job-spec)
+
+### Dependencies
```
I strongly feel that there is no need for this section

Kenny407 replied:
Removing it, I didn't find any usage in kubernetes or our current approach rather than just a list of applications

Olivier-Daire commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,193 @@
+Crons are tasks that run at a specific time or interval.
+You can use them to automate tasks, commonly as: reporting, backups, maintenance, emails and more.
+
+You can define the crons configuration with the following structure.
```
```suggestion
```

patrickportal commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,193 @@
+
+## Application Configuration
+
+In each file you will define the application that you want to deploy, few fields are required to setup a new application. There are many supported values which are optional that could be useful for your deployment. You can read more about them in the following sections and the schemas definition.
```
```suggestion
In each file you will define the application that you want to deploy, few fields are required to setup a new application. There are many optional supported values that could be useful for your deployment. You can read more about them in the following sections and the schemas definition.
```

patrickportal commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,193 @@
+  - name: process-name        # Process name - required - maxLength: 58
+    is_cronjob: false         # Defining a simple process
+    start_command: npm start  # Command executed when the pod is started
+# Defining a scheduled process
```
```suggestion
    # Defining a scheduled process
```

patrickportal commented on validator/docs/schemas.md:
```diff
@@ -0,0 +1,193 @@
+resources:
+  my-resource:  # Name of the resource to define
+    limit:
+      cpu:      # STRING / NUMBER
```
Why would it be a string for CPU? 🤔

Kenny407 replied:
`'50'`, 50 both are valid according to the official doc 😮

patrickportal commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,193 @@
+
+```yaml
+resources:
+  my-resource:  # Name of the resource to define
````
It should be the name or the process here, right?

patrickportal commented on validator/docs/schemas.md:
````diff
@@ -0,0 +1,193 @@
+```yaml
+resources:
+  my-resource:  # Name of the resource to define
+    limit:
````
I don't know what `limit` `limits` `request` `requests` keys stands for here, kubernetes things?

Kenny407 replied:
For simplicity I kept only limits and requests, and yes they are kubernetes specifics. I will add the link to the correct documentation.
